compiler/src/dotty/tools/dotc/semanticdb/TypeOps.scala

```diff
@@ -0,0 +1,358 @@
+      loop(tpe)
+
+    private def toSemanticType(using LinkMode, SemanticSymbolBuilder, Context): s.Type =
+      import ConstantOps._
+      def loop(tpe: Type): s.Type = tpe match {
+        case ExprType(tpe) =>
+          val stpe = loop(tpe)
+          s.ByNameType(stpe)
+
+        case TypeRef(pre, sym: Symbol) =>
+          val spre = if tpe.hasTrivialPrefix then s.Type.Empty else loop(pre)
+          val ssym = sym.symbolName
+          s.TypeRef(spre, ssym, Seq.empty)
+
+        case TermRef(pre, sym: Symbol) =>
+          val spre = if(tpe.hasTrivialPrefix) s.Type.Empty else loop(pre)
+          val ssym = sym.symbolName
+          s.SingleType(spre, ssym)
+
+        case tref: ParamRef =>
+          val key = (tref.binder, tref.paramName)
+          paramRefSymtab.get(key) match {
+            case Some(ref) =>
+              val ssym = ref.symbolName
+              tref match {
+                case _: TypeParamRef => s.TypeRef(s.Type.Empty, ssym, Seq.empty)
+                case _: TermParamRef => s.SingleType(s.Type.Empty, ssym)
+              }
+            case None => // shouldn't happen
+              s.Type.Empty
+          }
+
+        case ThisType(TypeRef(_, sym: Symbol)) =>
+          s.ThisType(sym.symbolName)
+
+        case SuperType(thistpe, supertpe) =>
+          val spre = loop(thistpe.typeSymbol.info)
+          val ssym = supertpe.typeSymbol.symbolName
+          s.SuperType(spre, ssym)
+
+        // val clazzOf = classOf[...]
+        case ConstantType(const) if const.tag == core.Constants.ClazzTag =>
+          loop(const.typeValue)
+
+        case ConstantType(const) =>
+          s.ConstantType(const.toSemanticConst)
+
+        case rt @ RefinedType(parent, name, info) =>
+          // `X { def x: Int; def y: Int }`
+          // RefinedType(
+          //   parent = RefinedType(
+          //     parent = TypeRef(..., X)
+          //     ...
+          //   )
+          //   refinedName = x
+          //   refinedInfo = TypeRef(..., Int)
+          // )
+          type RefinedInfo = (core.Names.Name, Type)
+          def flatten(tpe: Type, acc: List[RefinedInfo]): (Type, List[RefinedInfo]) = tpe match {
+            case RefinedType(parent, name, info) =>
+              flatten(parent, acc :+ (name, info))
+            case _ =>
+              (tpe, acc)
+          }
+
+          // flatten parent types to list
+          // e.g. `X with Y with Z { refined }`
+          // RefinedType(parent = AndType(X, AndType(Y, Z)), ...)
+          // => List(X, Y, Z)
+          def flattenParent(parent: Type): List[s.Type] = parent match {
+            case AndType(tp1, tp2) =>
+              flattenParent(tp1) ++ flattenParent(tp2)
+            case _ => List(loop(parent))
+          }
+
+          val (parent, refinedInfos) = flatten(rt, List.empty)
+          val stpe = s.IntersectionType(flattenParent(parent))
+
+          val decls = refinedInfos.flatMap { (name, _) =>
+            refinementSymtab.get((rt, name))
```

Comment:
same comment as for the symbol lookup for method parameters, we should produce some sort of notification when there is a missing key